src/outputs/histogram.cpp

```diff
@@ -0,0 +1,554 @@
+  const auto x_edge_min = hist.x_edge_min;
+  const auto x_edge_dbin = hist.x_edge_dbin;
+  const auto y_edge_min = hist.y_edge_min;
+  const auto y_edge_dbin = hist.y_edge_dbin;
+  const auto hist_ndim = hist.ndim;
+  const auto weight_by_vol = hist.weight_by_vol;
+  auto result = hist.result;
+  auto scatter = hist.scatter_result;
+
+  // Reset ScatterView from previous output
+  scatter.reset();
+  // Also reset the histogram from previous call.
+  // Currently still required for consistent results between host and device backends, see
+  // https://github.com/kokkos/kokkos/issues/6363
+  Kokkos::deep_copy(result, 0);
+
+  const int num_partitions = pm->DefaultNumPartitions();
+
+  for (int p = 0; p < num_partitions; p++) {
+    auto &md = pm->mesh_data.GetOrAdd("base", p);
+
+    const auto x_var_pack_string = x_var_type == VarType::Var
+                                       ? std::vector<std::string>{hist.x_var_name}
+                                       : std::vector<std::string>{};
+    const auto x_var = md->PackVariables(x_var_pack_string);
+
+    const auto y_var_pack_string = y_var_type == VarType::Var
+                                       ? std::vector<std::string>{hist.y_var_name}
+                                       : std::vector<std::string>{};
+    const auto y_var = md->PackVariables(y_var_pack_string);
+
+    const auto binned_var_pack_string =
+        binned_var_component == -1 ? std::vector<std::string>{}
+                                   : std::vector<std::string>{hist.binned_var_name};
+    const auto binned_var = md->PackVariables(binned_var_pack_string);
+
+    const auto weight_var_pack_string =
+        weight_var_component == -1 ? std::vector<std::string>{}
+                                   : std::vector<std::string>{hist.weight_var_name};
+    const auto weight_var = md->PackVariables(weight_var_pack_string);
+
+    const auto ib = md->GetBoundsI(IndexDomain::interior);
+    const auto jb = md->GetBoundsJ(IndexDomain::interior);
+    const auto kb = md->GetBoundsK(IndexDomain::interior);
+
+    parthenon::par_for(
+        DEFAULT_LOOP_PATTERN, "CalcHist", DevExecSpace(), 0, md->NumBlocks() - 1, kb.s,
+        kb.e, jb.s, jb.e, ib.s, ib.e,
+        KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+          auto &coords = x_var.GetCoords(b);
+          auto x_val = std::numeric_limits<Real>::quiet_NaN();
+          if (x_var_type == VarType::X1) {
+            x_val = coords.Xc<1>(k, j, i);
+          } else if (x_var_type == VarType::X2) {
+            x_val = coords.Xc<2>(k, j, i);
+          } else if (x_var_type == VarType::X3) {
+            x_val = coords.Xc<3>(k, j, i);
+          } else if (x_var_type == VarType::R) {
+            x_val = Kokkos::sqrt(SQR(coords.Xc<1>(k, j, i)) + SQR(coords.Xc<2>(k, j, i)) +
+                                 SQR(coords.Xc<3>(k, j, i)));
+          } else {
+            x_val = x_var(b, x_var_component, k, j, i);
```

Comment:
worth noting that (I think?) this even works for tensor components, since you're using a pack, and that will flatten them out...